src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyExecutionManager.cs

```diff
@@ -96,7 +98,10 @@ public virtual int StartTestRun(TestRunCriteria testRunCriteria, ITestRunEventsH
                     testSources = testRunCriteria.Tests.GroupBy(tc => tc.Source).Select(g => g.Key);
                 }
 
-                this.isCommunicationEstablished = this.SetupChannel(testSources, this.cancellationTokenSource.Token);
+                lock (locObject)
```

Comment:
In case when the test request timed out, before the set up channel could complete, you are allowing the testhost launch to complete, & wait for the communication to establish, & then you will send the cancellation request. Would it not be helpful if you just set cancellation token & hope that if the test host was not launched, then it might get cancel, & lock() the Cancel request?